app/src/cose.h

```diff
@@ -20,87 +20,186 @@
+    bool is_critical(const std::variant<int64_t, std::string>& label) const
+    {
+      if (!crit.has_value())
+      {
+        return false;
+      }
+      for (const auto& crit_label : crit.value())
+      {
+        if (crit_label == label)
+        {
+          return true;
+        }
+      }
+      return false;
+    }
+  };
 
-    QCBORError qcbor_result;
+  struct UnprotectedHeader
+  {
+    // We currently expect only notary to use the unprotected header and
+    // we expect to find only the x5chain in there.
+    std::optional<std::vector<std::vector<uint8_t>>> x5chain;
+  };
 
-    QCBORDecodeContext ctx;
-    QCBORDecode_Init(
-      &ctx, cbor::from_bytes(cose_sign1), QCBOR_DECODE_MODE_NORMAL);
+  std::vector<std::vector<uint8_t>> decode_x5chain(
+    QCBORDecodeContext& ctx, const QCBORItem& x5chain)
+  {
+    std::vector<std::vector<uint8_t>> parsed;
 
-    QCBORDecode_EnterArray(&ctx, nullptr);
-    qcbor_result = QCBORDecode_GetError(&ctx);
-    if (qcbor_result != QCBOR_SUCCESS)
+    if (x5chain.uDataType == QCBOR_TYPE_ARRAY)
     {
-      throw COSEDecodeError("Failed to parse COSE_Sign1 outer array");
+      QCBORDecode_EnterArrayFromMapN(&ctx, COSE_HEADER_PARAM_X5CHAIN);
+      while (true)
+      {
+        QCBORItem item;
+        auto result = QCBORDecode_GetNext(&ctx, &item);
+        if (result == QCBOR_ERR_NO_MORE_ITEMS)
+        {
+          break;
+        }
+        if (result != QCBOR_SUCCESS)
+        {
+          throw COSEDecodeError("Item in x5chain is not well-formed.");
+        }
+        if (item.uDataType == QCBOR_TYPE_BYTE_STRING)
+        {
+          parsed.push_back(cbor::as_vector(item.val.string));
+        }
+        else
+        {
+          throw COSEDecodeError(
+            "Next item in x5chain was not of type byte string.");
+        }
+      }
+      QCBORDecode_ExitArray(&ctx);
+      if (parsed.empty())
+      {
+        throw COSEDecodeError("x5chain array length was 0 in COSE header.");
+      }
     }
-
-    uint64_t tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
-    if (tag != CBOR_TAG_COSE_SIGN1)
+    else if (x5chain.uDataType == QCBOR_TYPE_BYTE_STRING)
     {
-      throw COSEDecodeError("COSE_Sign1 is not tagged");
+      parsed.push_back(cbor::as_vector(x5chain.val.string));
+    }
+    else
+    {
+      CCF_APP_FAIL("Type: {}", x5chain.uDataType);
+      throw COSEDecodeError(
+        "Value type of x5chain in COSE header is not array or byte "
+        "string.");
     }
 
+    return parsed;
+  }
+
+  ProtectedHeader decode_protected_header(QCBORDecodeContext& ctx)
+  {
+    ProtectedHeader parsed;
+
+    // Adapted from parse_cose_header_parameters in t_cose_parameters.c.
+    // t_cose doesn't support custom header parameters yet.
+
+    QCBORError qcbor_result;
+
     struct q_useful_buf_c protected_parameters;
     QCBORDecode_EnterBstrWrapped(
       &ctx, QCBOR_TAG_REQUIREMENT_NOT_A_TAG, &protected_parameters);
```

Comment:
According to the QCBOR docs, we should be able to pass NULL as the third parameter and not need to define `protected_parameters` (since we don't use it anyway).

Reply:
Done in https://github.com/microsoft/scitt-ccf-ledger/pull/90